endpoint: moveNewFilesTo performance and error handling improvements

learnitall opened the pull request:
These commits make changes to the function moveNewFilesTo to reduce its complexity from O(n^2) to O(n), decrease its CPU and memory usage, and improve its error handling. The function moveNewFilesTo is a utility function that may be called in the endpoint regeneration hot path.

Details of the changes can be found in each commit.

Please ensure your pull request adheres to the following guidelines:

- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Thanks for contributing!

I believe this needs the following labels:

* `release-note/misc`
* `kind/enhancement`

Also, I think this could be backported, but I haven't taken the time yet to dig into how feasible this would be.

learnitall committed:
endpoint: Reduce CPU and memory usage of function moveNewFilesTo

This commit makes changes to the function moveNewFilesTo to reduce its
complexity from O(n^2) to O(n) and decrease its CPU and memory usage.
The function moveNewFilesTo is a utility function that may be called
in the endpoint regeneration hot path. This commit makes the
following changes to speed it up and reduce memory allocations:

* Use a map to look up if a file from the source directory exists in
  the destination directory. Before this commit, this lookup was done
  using a nested for loop, looping over slices that contain files
  in the source and destination directories.
* Use `os.File.Readdirnames` instead of os.ReadDir. The function
  `os.ReadDir` sorts the slice of returned files, which is not needed
  for this use case. Additionally, Readdirnames returns a string slice
  of file names, rather than a slice of `os.DirEntry` structs, meaning
  the name of each file does need to be accessed using an accessor
  method.

The benefits of these changes will pay off in scale. The variables
impacting the resource consumption of this function are:

* The number of files to move between two the directories.
* Rate of endpoint regeneration, which is impacted by policy churn,
  pod churn, node churn, label updates, among others.

Even if these changes save 0.1% of CPU time, this adds up to a core
of savings just by creating 1000 pods (10 pods per node in a 100 node
cluster).

In the future, more work can be done to look at the higher-level
context of how an endpoint's eBPF programs are regenerated, to find
greater optimizations.

This commit additionally introduces benchmark tests. Here are results
comparing the implementation of moveNewFilesTo before and after this
commit using benchstat. Each benchmark was executed 10 times.

```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                   │ MoveNewFilesTo1.txt │     MoveNewFilesTo1-after.txt      │
                   │       sec/op        │   sec/op     vs base               │
MoveNewFilesTo1-20           5.583µ ± 1%   5.514µ ± 1%  -1.24% (p=0.000 n=10)

                   │ MoveNewFilesTo1.txt │     MoveNewFilesTo1-after.txt      │
                   │        B/op         │    B/op     vs base                │
MoveNewFilesTo1-20            486.0 ± 0%   373.0 ± 0%  -23.25% (p=0.000 n=10)

                   │ MoveNewFilesTo1.txt │     MoveNewFilesTo1-after.txt      │
                   │      allocs/op      │ allocs/op   vs base                │
MoveNewFilesTo1-20            13.00 ± 0%   10.00 ± 0%  -23.08% (p=0.000 n=10)
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                   │ MoveNewFilesTo5.txt │     MoveNewFilesTo5-after.txt      │
                   │       sec/op        │   sec/op     vs base               │
MoveNewFilesTo5-20           6.878µ ± 1%   6.383µ ± 1%  -7.20% (p=0.000 n=10)

                   │ MoveNewFilesTo5.txt │     MoveNewFilesTo5-after.txt      │
                   │        B/op         │    B/op     vs base                │
MoveNewFilesTo5-20           1137.0 ± 0%   641.0 ± 0%  -43.62% (p=0.000 n=10)

                   │ MoveNewFilesTo5.txt │     MoveNewFilesTo5-after.txt      │
                   │      allocs/op      │ allocs/op   vs base                │
MoveNewFilesTo5-20            29.00 ± 0%   19.00 ± 0%  -34.48% (p=0.000 n=10)
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                    │ MoveNewFilesTo10.txt │     MoveNewFilesTo10-after.txt      │
                    │        sec/op        │   sec/op     vs base                │
MoveNewFilesTo10-20            9.173µ ± 1%   7.644µ ± 1%  -16.67% (p=0.000 n=10)

                    │ MoveNewFilesTo10.txt │      MoveNewFilesTo10-after.txt      │
                    │         B/op         │     B/op      vs base                │
MoveNewFilesTo10-20           1.992Ki ± 0%   1.322Ki ± 0%  -33.63% (p=0.000 n=10)

                    │ MoveNewFilesTo10.txt │     MoveNewFilesTo10-after.txt     │
                    │      allocs/op       │ allocs/op   vs base                │
MoveNewFilesTo10-20             48.00 ± 0%   30.00 ± 0%  -37.50% (p=0.000 n=10)
```

The use of goroutines was purposefully omitted from these changes for
two reasons:

1. The expected number of files in the source directory given to
   moveNewFilesTo is four. The max number of extra goroutines that
   could be expected on a given machine running Cilium is four times
   the number of endpoints on the node. The impact of these extra
   goroutines is difficult to predict and debated.
2. Using goroutines to parallelize renames didn't make a substantial
   difference in the amount of CPU time saved, while increasing overall
   memory usage due to the overhead of goroutine creation.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

christarazi commented on pkg/endpoint/directory.go:
```diff
@@ -48,27 +49,48 @@ func (e *Endpoint) backupDirectoryPath() string {
+			go func() {
+				defer wg.Done()
+				os.Rename(filepath.Join(oldDir, oldFile), filepath.Join(newDir, oldFile))
+			}()
```
I'm not sure how many `oldFiles` there can be but this is spinning off as many goroutines as there are `oldFiles`. While the body of the goroutine is quite small--just renaming a file, this can put additional pressure onto the Go runtime if it is already managing many goroutines or if it is already under CPU pressure. We probably want to bound this number of goroutines to at least the number of CPUs, but even then this code is running for each endpoint that's regenerating, which can happen in parallel.

marseel replied:
I agree it would be nice to understand how many `oldFiles` there can be. I don't think the number of goroutines matters that much tbh, but we can limit concurrency here for sure - even something constant should be fine, like 5 or 10, as this is mostly io operation. Also, it would be nice to handle error from `os.Rename` as we haven't done that before...

learnitall replied:
I pushed some changes to use an `errgroup.Group` to limit the number of goroutines to five and allow for us to bubble-up errors from `os.Rename`. Five goroutines is a small enough amount that it won't have a drastic impact on the runtime, and since these operations are IO bound we don't need to worry about these goroutines contributing to CPU pressure.

@christarazi helped me look into the number of `oldFiles` that we could expect and it should stay less than 10. There's four static files that are present here with an additional three that could be created if debug mode is enabled.

Hey @aanm, it looks like you originally authored this function five years ago in https://github.com/cilium/cilium/commit/f6c4385a43943c7be71bd8ca3a0582cab24db20b. Do you remember if there was a reason for not checking the error on `os.Rename`?

aanm replied:
@learnitall no specific reason.

aanm replied:
I didn't look to much into this PR but from what I can see the total number of go routines will depend on the number of endpoints on the node times ~4 (number of old files).

learnitall replied:
I did a quick test to see what would happen if we kept the other changes in this PR, but run `os.Rename` in serial, and I don't see a difference. The sec/op stays about the same, and memory usage looks better since we skip creating the goroutines:

```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                   │ MoveNewFilesTo1-before--.txt │   MoveNewFilesTo1-after.txt   │
                   │            sec/op            │   sec/op     vs base          │
MoveNewFilesTo1-20                    5.508µ ± 0%   5.514µ ± 1%  ~ (p=0.363 n=10)

                   │ MoveNewFilesTo1-before--.txt │     MoveNewFilesTo1-after.txt      │
                   │             B/op             │    B/op     vs base                │
MoveNewFilesTo1-20                     438.0 ± 0%   373.0 ± 0%  -14.84% (p=0.000 n=10)

                   │ MoveNewFilesTo1-before--.txt │     MoveNewFilesTo1-after.txt     │
                   │          allocs/op           │ allocs/op   vs base               │
MoveNewFilesTo1-20                     11.00 ± 0%   10.00 ± 0%  -9.09% (p=0.000 n=10)
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                   │ MoveNewFilesTo5-before--.txt │   MoveNewFilesTo5-after.txt   │
                   │            sec/op            │   sec/op     vs base          │
MoveNewFilesTo5-20                    6.376µ ± 1%   6.383µ ± 1%  ~ (p=0.897 n=10)

                   │ MoveNewFilesTo5-before--.txt │     MoveNewFilesTo5-after.txt     │
                   │             B/op             │    B/op     vs base               │
MoveNewFilesTo5-20                     705.0 ± 0%   641.0 ± 0%  -9.08% (p=0.000 n=10)

                   │ MoveNewFilesTo5-before--.txt │     MoveNewFilesTo5-after.txt     │
                   │          allocs/op           │ allocs/op   vs base               │
MoveNewFilesTo5-20                     20.00 ± 0%   19.00 ± 0%  -5.00% (p=0.000 n=10)
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                    │ MoveNewFilesTo10-before--.txt │  MoveNewFilesTo10-after.txt   │
                    │            sec/op             │   sec/op     vs base          │
MoveNewFilesTo10-20                     7.694µ ± 4%   7.644µ ± 1%  ~ (p=0.089 n=10)

                    │ MoveNewFilesTo10-before--.txt │     MoveNewFilesTo10-after.txt      │
                    │             B/op              │     B/op      vs base               │
MoveNewFilesTo10-20                    1.385Ki ± 0%   1.322Ki ± 0%  -4.51% (p=0.000 n=10)

                    │ MoveNewFilesTo10-before--.txt │    MoveNewFilesTo10-after.txt     │
                    │           allocs/op           │ allocs/op   vs base               │
MoveNewFilesTo10-20                      31.00 ± 0%   30.00 ± 0%  -3.23% (p=0.000 n=10)```

learnitall replied:
Given the above, I'll go ahead and remove usage of goroutines to make this PR safer.

christarazi commented on pkg/endpoint/directory.go:
```diff
@@ -48,27 +48,43 @@ func (e *Endpoint) backupDirectoryPath() string {
+	newFilesHash := make(map[string]bool, len(newFiles))
+	for _, f := range newFiles {
+		newFilesHash[f] = true
+	}
```
`newFilesHash` can use `struct{}` as a value instead of `bool` since `true` is the only value that's ever set in the map.

learnitall replied:
Ah good catch, changed.

christarazi commented on pkg/endpoint/directory.go:
```diff
@@ -81,7 +81,9 @@ func moveNewFilesTo(oldDir, newDir string) error {
 		if ok = newFilesHash[oldFile]; !ok {
-			os.Rename(filepath.Join(oldDir, oldFile), filepath.Join(newDir, oldFile))
+			if err := os.Rename(filepath.Join(oldDir, oldFile), filepath.Join(newDir, oldFile)); err != nil {
+				return err
```
Could you add some context here like `fmt.Errorf("failed to rename new endpoint state files: %w", err)`

learnitall replied:
Done

learnitall committed:
endpoint: Improve error verbosity in moveNewFilesTo

This commit modifies moveNewFilesTo return with an error if
one occurs while renaming a file. Additionally, extra context
has been added to errors that may occur when opening and
listing directories. This allows for these errors to be more
visible, which may help in debugging and troubleshooting.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

christarazi reviewed (approved):
LGTM thanks. There are some CI failures including lint failures.
